lib/spack/spack/test/cmd_extensions.py

```diff
@@ -79,33 +107,150 @@ def hello(parser, args):
+    yield spack.main.SpackCommand('hello')
 
 
 def test_simple_command_extension(hello_world_cmd):
+    """Basic test of a functioning command."""
     output = hello_world_cmd()
     assert 'Hello world!' in output
 
 
+def test_multi_extension_search(hello_world_extension, tmpdir):
+    """Ensure we can find an extension command even if it's not in the first
+    place we look.
+    """
+
+    extra_ext_name = 'testcommand2'
+    extra_ext = Extension(extra_ext_name,
+                          tmpdir.mkdir('spack-' + extra_ext_name))
+    with spack.config.override('config:extensions', [str(extra_ext.root)]):
+        assert ('Hello world') in spack.main.SpackCommand('hello-world')()
+
+
+def test_duplicate_module_load(hello_world_cmd):
+    """Ensure duplicate module load attempts are successful.
+
+    The command module will already have been loaded once by the
+    hello_world_cmd fixture.
+    """
+    assert ('Hello world') in spack.main.SpackCommand('hello-world')()
+
+
+@pytest.mark.parametrize('extension',
+                         ['testcommand', 'hyphenated-extension'],
+                         ids=['simple', 'hyphenated_extension_name'],
+                         indirect=True)
 def test_command_with_import(hello_world_with_module_in_root):
+    """Ensure we can write a functioning command with multiple imported
+    subcommands, including where the extension name contains a hyphen.
+    """
     output = hello_world_with_module_in_root('world')
     assert 'Hello world!' in output
     output = hello_world_with_module_in_root('folks')
     assert 'Hello folks!' in output
     output = hello_world_with_module_in_root('global')
     assert 'bar' in output
+
+
+def test_missing_command():
+    """Ensure that we raise the expected exception if the desired command is
+    not present.
+    """
+    with pytest.raises(spack.extensions.CommandNotFoundError):
+        spack.cmd.get_module("no-such-command")
+
+
+@pytest.mark.parametrize('extension_data',
+                         [('/my/bad/extension', False),
+                          ('', False),
+                          ('/my/bad/spack--extra-hyphen', False),
+                          ('/my/good/spack-extension', True),
+                          ('/my/still/good/spack-extension/', True),
+                          ('/my/spack-hyphenated-extension', True)],
+                         ids=['no_stem', 'vacuous', 'leading_hyphen',
+                              'basic_good', 'trailing_slash', 'hyphenated'])
+def test_extension_naming(extension_data, config):
+    """Ensure that we are correctly validating configured extension paths
+    for conformity with the rules: the basename should match
+    ``spack-<name>``; name may have embedded extensions but not begin
+    with one.
+    """
+    ext_path = extension_data[0]
+    expected_exception\
+        = spack.extensions.CommandNotFoundError if \
+        extension_data[1] else \
+        spack.extensions.ExtensionNamingError
```

Comment:
@alalazo I understand that it can, but I'm having trouble understanding why that is preferable. I've made it work, but because of the aforementioned recommendation for full scoping of exceptions, saving four lines of body has added six lines to the parameterization directive. I can commit this anyway, but I'd really like to understand why you think it's preferable, first.

Reply:
> I've made it work, but because of the aforementioned recommendation for full scoping of exceptions, saving four lines of body has added six lines to the parameterization directive

Line of codes are just a loose proxy for code complexity, which is not increased here - there are more lines just because of an arbitrary choice to keep them at 80 chars. In this case:

1.  `expected_exception` conveys more meaning than ` a if expected_data[1] else b`
2. The test is already set up to accept a new type of exception by just adding a new parameter

Reply:
Fair enough---done.